File: OOP_Encapsulation.cpp

// class
// thuoc tinh:attribute
// phuong thuc : behavior ,  method
// object: doi tuong

// Accesmodifier :
//  private : chi truy cap trong cung class , khong the truy cap ben ngoai
// public : tat ca moi noi
// protected : cung class, subclass ke thua
// static : khong khoi tao ,dung chung cho ca class
//friend function,friend class

#include <bits/stdc++.h>
#include <iostream>
#include <iomanip>
#include <vector>
#include <algorithm>
#include <string>

class SinhVien {
private:
    static int dem;
    std::string maSV;
    std::string id;
    std::string name;
    std::string birthday;
    double gpa;

    std::string sinhMaSV() {
        return "SV" + std::string(3 - std::to_string(dem).length(), '0') + std::to_string(dem);
    }

public:
    SinhVien() {
        ++dem;
        maSV = sinhMaSV();
    }

    friend void ChuanHoa(SinhVien&);

    // Destructor (hiện không cần làm gì)
    ~SinhVien() {}

    // Hàm nhập thông tin từ người dùng
    void nhap() {
        std::cout << "Nhap id: ";
        std::cin >> id;
        std::cin.ignore();
        std::cout << "Nhap ten: ";
        std::getline(std::cin, name);
        std::cout << "Nhap ngay sinh: ";
        std::cin >> birthday;
        std::cout << "Nhap GPA: ";
        std::cin >> gpa;
    }

    // Hàm in thông tin sinh viên
    void in() const {
        std::cout << maSV << " " << id << " " << name << " " << birthday << " "
                  << std::fixed << std::setprecision(2) << gpa << '\n';
    }

    double getGPA() const {
        return gpa;
    }
};
void ChuanHoa(SinhVien &a) {
    std::string res = "";
    std::stringstream ss(a.name); 
    std::string token; 

    while (ss >> token) {
        res+= toupper(token[0]); 
        for(int i = 1; i < token.length();i++) {
            res+=tolower(token[i]); 
        }
        res+= " "; 
    }
    res.erase (res.length() - 1) ; 
    a.name = res ; 
}
// Khởi tạo biến static
int SinhVien::dem = 0;

// Hàm so sánh GPA giảm dần
bool soSanhGPA(const SinhVien& a, const SinhVien& b) {
    return a.getGPA() > b.getGPA();
}

// Chương trình chính
int main() {
    char tiepTuc = 'y';
    while (tiepTuc == 'y' || tiepTuc == 'Y') {
        int n;
        std::cout << "Nhap so sinh vien: ";
        std::cin >> n;

        std::vector<SinhVien> danhSach;
        for (int i = 0; i < n; ++i) {
            std::cout << "Nhap sinh vien thu " << i + 1 << ":\n";
            SinhVien sv;
            sv.nhap();
            ChuanHoa(sv); 
            danhSach.push_back(sv);
        }

        std::sort(danhSach.begin(), danhSach.end(), soSanhGPA);

        std::cout << "\nDanh sach sinh vien theo GPA giam dan:\n";
        for (const auto& sv : danhSach) {
            sv.in();
        }

        std::cout << "\nBan co muon tiep tuc? (y/n): ";
        std::cin >> tiepTuc;
        std::cin.ignore();
    }

    return 0;
}

File: OOP_Encapsulation.java
import java.util.Scanner;

class SinhVien {
    private String id, name, birthday;
    private double gpa;
    String maSv;

@Override
public String toString() {
    return maSv + " " + id +" " + name + " " + birthday + " " + gpa; 
}

    static int dem; 
    public SinhVien() {
        maSv = String.format("SV%03d",++dem); // "%03d" = 3 chữ số, thêm số 0 ở đầu nếu thiếu
    }

    public SinhVien(String id, String name, String birthday, double gpa) {
        this.id = id;
        this.name = name;
        this.birthday = birthday;
        this.gpa = gpa;
    }

    public double getGPA() {
        return gpa;
    }

    public void nhap() {
        Scanner sc = new Scanner(System.in);

        System.out.print("Nhap id: ");
        id = sc.nextLine();
        System.out.print("Nhap name: ");
        name = sc.nextLine();
        System.out.print("Nhap ngay sinh: ");
        birthday = sc.nextLine();
        System.out.print("Nhap GPA: ");
        gpa = sc.nextDouble();
        sc.nextLine(); // Clear buffer
    }

    // public void in() {
    //     System.out.println(maSv+ " " + id + " " + name + " " + birthday + " " + gpa);
    // }

    // public static boolean Compare(SinhVien a, SinhVien b) {
    //     return a.getGPA() > b.getGPA();
    // }
}



public class OOP_Encapsulation {
    public static void main(String[] args) {
     
        int n;
        Scanner sc = new Scanner(System.in);
        System.out.print("Nhap so sinh vien: ");
        n = sc.nextInt();
        sc.nextLine(); // clear buffer

        SinhVien[] a = new SinhVien[100];

        for (int i = 0; i < n; i++) {
            a[i] = new SinhVien();
            a[i].nhap();
        }
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (a[i].getGPA() < a[j].getGPA()) {
                    SinhVien temp;
                    temp = a[i];
                    a[i] = a[j];
                    a[j] = temp;
                }
            }
        }

        System.out.println("\nDanh sach sinh vien:");
        for (int i = 0; i < n; i++) {
            System.out.println(a[i]);
        }
    }
}
